### PR TITLE
Supporting Sigma v4.0.1 signing functions 

### PR DESCRIPTION
**Pull request by dav009**
https://github.com/ergoplatform/ergo-appkit/issues/65

👋  ErgoTeam,  I took on this issue to get more familiar with the code base.  (tagged as `Good first issue`)
If the code in this PR makes sense I am happy to continue adding the needed unit-tests.

The API side of these changes are made in other PR on the ergo repo: https://github.com/ergoplatform/ergo/pull/1484/files

@aslesarenko I was also wondering if passing a SigmaBoolean makes sense at all. I was thinking on just passing the public key i.e:

```scala
 // no tx here
  def signMessage(message:  Array[Byte], hintsBag: HintsBag): Array[Byte] = {
     pk = _prover.publicKeys.head 
    _prover.signMessage(pk, message, hintsBag).getOrThrow
  }
```

----
- [x] unit tests

### CODE REVIEW

File: lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java [hunk 1]
```diff
@@ -63,8 +66,12 @@
      */
     SignedTransaction sign(UnsignedTransaction tx, int baseCost);
 
+    byte[] signMessage(SigmaBoolean sigmaTree, byte[] message, HintsBag hintsBag);
+
     ReducedTransaction reduce(UnsignedTransaction tx, int baseCost);
 
     SignedTransaction signReduced(ReducedTransaction tx, int baseCost);
+
+    boolean verifySignature(SigmaBoolean tx, byte[] message, byte[] signedMessage);
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java:75]:
```suggestion
    boolean verifySignature(SigmaBoolean sigmaTree, byte[] message, byte[] signedMessage);
```

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ErgoProverImpl.scala [hunk 2]
```diff
@@ -46,6 +48,11 @@ class ErgoProverImpl(_ctx: BlockchainContextBase,
     new SignedTransactionImpl(_ctx, signed, cost)
   }
 
+  def signMessage(sigmaTree: SigmaBoolean, message:  Array[Byte], hintsBag: HintsBag): Array[Byte] = {
+    _prover.signMessage(sigmaTree, message, hintsBag).getOrThrow
+  }
+
+
```
Comment (aslesarenko) [inline lib-impl/src/main/java/org/ergoplatform/appkit/impl/ErgoProverImpl.scala:55]:
```suggestion
```
Please remove unnecessary blank lines.

**Comment by aslesarenko** [comment]
@dav009 Yes, SigmaBoolean is a generalized public key so the code looks good, modulo comments. 
You can also add an overload of the methods which takes P2PKAddress.
Note, P2PKAddress can be either constructed directly or obtained for org.ergoplatform.appkit.Address.

Please go ahead with this PR, everything looks good.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

**Comment by dav009** [comment]
@aslesarenko added some unit tests . 
CI is still failing when publishing the snapshot artifacts 🤔 .

File: lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java [hunk 3]
```diff
@@ -63,8 +66,16 @@
      */
     SignedTransaction sign(UnsignedTransaction tx, int baseCost);
 
+    byte[] signMessage(SigmaBoolean sigmaTree, byte[] message, HintsBag hintsBag);
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java:69]: Every public method need full documentation including: 1) description 2) all parameter 3) return value. 
In general, the description shouldn't just repeat what is obvious from the signature, instead it should complement it with special considerations derived from the essence of the method (something you would want to read, it is were not your method)

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java [hunk 4]
```diff
@@ -66,16 +66,50 @@
...
+     * @param hintsBag additional hints for a signer (useful for distributed signing)
+     * @return signed message
+     */
     byte[] signMessage(P2PKAddress addr, byte[] message, HintsBag hintsBag);
 
     ReducedTransaction reduce(UnsignedTransaction tx, int baseCost);
 
     SignedTransaction signReduced(ReducedTransaction tx, int baseCost);
 
+     /**
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java:94]:
```suggestion
    /**
```

File: lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java [hunk 5]
```diff
@@ -66,16 +66,50 @@
...
+     *
+     * @param sigmaTree public key (represented as a tree)
+     * @param message message to verify
+     * @param signedMessage signature for the message
+     * @return whether signature is valid or not
+     */
     boolean verifySignature(SigmaBoolean sigmaTree, byte[] message, byte[] signedMessage);
-    
+
+     /**
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/ErgoProver.java:104]:
```suggestion
    /**
```

**Review by aslesarenko — APPROVED** [review]

**Comment by aslesarenko** [comment]
Now looks good, waiting for tests

**Comment by dav009** [comment]
@aslesarenko  added a round trip sign/verification.

**Pull request merged**